### ISSUE TITLE
0.0.74 __slots__ enforcement rejects stub-declared attribute assignments — breaks every protobuf _pb2.pyi

### ISSUE THREAD

**Issue by RKumsher** [astral-sh/ty#4375]
## Summary

Since 0.0.74 (`__slots__` support, astral-sh/ruff#27730), assigning to an attribute that a stub **declares** as a bare class-body annotation is rejected with `unresolved-attribute` when the stub also declares `__slots__ = ()`. This is the exact shape `protoc` emits in every generated `_pb2.pyi`, so any project that type-checks protobuf field assignments fails on every `message.field = value` under 0.0.74.

## Minimal repro (no protobuf involved)

```
# pkg/__init__.py    (empty)
# pkg/mod.py
class Plain:
    pass

# pkg/mod.pyi
class Plain:
    __slots__ = ()
    field: str
    def __init__(self, field: str = ...) -> None: ...

# main.py
from pkg.mod import Plain

p = Plain()
p.field = "x"
```

```
$ uvx ty@0.0.73 check .
All checks passed!

$ uvx ty@0.0.74 check .
error[unresolved-attribute]: Unresolved attribute `field` on type `Plain`
 --> main.py:4:1
  |
4 | p.field = "x"
  | ^^^^^^^
```

## Real-world case: protobuf

`protoc`'s Python stubs universally declare `__slots__ = ()` on message classes and expose fields as bare annotations. For any proto like:

```proto
message Article {
  string title = 1;
}
```

the generated `_pb2.pyi` contains:

```pyi
class Article(_message.Message):
    __slots__ = ()
    TITLE_FIELD_NUMBER: _ClassVar[int]
    title: str
    def __init__(self, title: _Optional[str] = ...) -> None: ...
```

At runtime the fields are metaclass-provided properties, so `article.title = "..."` succeeds — the empty `__slots__` reflects "no user-defined instance attributes," not "fields are unwritable." Under 0.0.74 every such assignment is flagged. Bisected cleanly: 0.0.73 passes, 0.0.74 fails, identical environments.

## Expected

mypy and pyright both treat a bare class-body annotation in a stub as a writable attribute declaration regardless of `__slots__` membership, which is why this stub idiom has been invisible until now. Ideally ty matches that for stub files (a stub author writing `field: str` is asserting the attribute exists and is settable — the stub has no way to express the property machinery protoc's metaclass provides), or at minimum the protobuf story gets documented guidance.

Related: astral-sh/ruff#27969 (in flight) improves the diagnostic for exactly this "declared but not included in `__slots__`" case — but the protobuf ecosystem needs the write to be *accepted* in stubs, not better-explained.

Happy to provide more detail — thanks for ty!

**Comment by AlexWaygood** [comment]
Thanks for the report! Interesting, I don't think we knew that protobuf stubs generally made use of this pattern.

> the stub has no way to express the property machinery protoc's metaclass provides

I'm not sure that's true? Can't the stub do something like this? ty understands that `field` is a writable attribute here -- it has a very good understanding of how descriptors work:

```py
class Plain:
    __slots__ = ()
    @property
    def field(self) -> str: ...
    @field.setter
    def field(self, value: str) -> None: ...
    def __init__(self, field: str = "") -> None: ...
```

We should probably move this diagnostic into a dedicated error code, at least, however, so that users who are especially affected by this can turn it off without having to turn off all `unresolved-attribute` errors.

**Comment by RKumsher** [comment]
Thanks, fair point, and good to know ty models the property shape correctly. The problem is these stubs are protoc-generated (`--pyi_out` / buf's protocolbuffers/pyi plugin), so downstream projects can't practically rewrite them.  The property shape is really something protoc's generator should emit (scalar fields only; message-typed fields aren't assignable at runtime anyway, which the current bare annotations also get wrong). Until something lands there, a dedicated error code we can disable narrowly would be perfect. Happy to test a nightly against real protoc output.
